serde_v8/src/ser.rs

```diff
@@ -0,0 +1,649 @@
+  }
+
+  /// For compatibility with serde-json, serialises unit variants as "Variant" strings.
+  fn serialize_unit_variant(
+    self,
+    _name: &'static str,
+    _variant_index: u32,
+    variant: &'static str,
+  ) -> JsResult<'a> {
+    Ok(v8_struct_key(&mut self.scope.borrow_mut(), variant).into())
+  }
+
+  fn serialize_newtype_struct<T: ?Sized + Serialize>(
+    self,
+    _name: &'static str,
+    value: &T,
+  ) -> JsResult<'a> {
+    value.serialize(self)
+  }
+
+  fn serialize_newtype_variant<T: ?Sized + Serialize>(
+    self,
+    _name: &'static str,
+    _variant_index: u32,
+    variant: &'static str,
+    value: &T,
+  ) -> JsResult<'a> {
+    let scope = self.scope.clone();
+    let x = self.serialize_newtype_struct(variant, value)?;
+    VariantSerializer::new(scope, variant, x).end(Ok)
+  }
+
+  /// Serialises any Rust iterable into a JS Array
+  fn serialize_seq(self, _len: Option<usize>) -> Result<Self::SerializeSeq> {
+    Ok(ArraySerializer::new(self.scope))
+  }
+
+  fn serialize_tuple(self, len: usize) -> Result<Self::SerializeTuple> {
+    self.serialize_seq(Some(len))
+  }
+
+  fn serialize_tuple_struct(
+    self,
+    _name: &'static str,
+    len: usize,
+  ) -> Result<Self::SerializeTupleStruct> {
+    self.serialize_tuple(len)
+  }
+
+  fn serialize_tuple_variant(
+    self,
+    _name: &'static str,
+    _variant_index: u32,
+    variant: &'static str,
+    len: usize,
+  ) -> Result<Self::SerializeTupleVariant> {
+    Ok(VariantSerializer::new(
+      self.scope.clone(),
+      variant,
+      self.serialize_tuple_struct(variant, len)?,
+    ))
+  }
+
+  fn serialize_map(self, _len: Option<usize>) -> Result<Self::SerializeMap> {
+    // Serializes a rust Map (e.g: BTreeMap, HashMap) to a v8 Object
+    // TODO: consider allowing serializing to v8 Maps (e.g: via a magic type)
+    // since they're lighter and better suited for K/V data
+    // and maybe restrict keys (e.g: strings and numbers)
+    Ok(MapSerializer::new(self.scope))
+  }
+
+  /// Serialises Rust typed structs into plain JS objects.
+  fn serialize_struct(
+    self,
+    name: &'static str,
+    _len: usize,
+  ) -> Result<Self::SerializeStruct> {
+    if name == magic::NAME {
+      let m = MagicSerializer {
+        scope: self.scope,
+        v8_value: None,
+      };
+      return Ok(StructSerializers::Magic(m));
+    }
+    let o = ObjectSerializer::new(self.scope);
+    Ok(StructSerializers::Regular(o))
+  }
+
+  fn serialize_struct_variant(
+    self,
+    _name: &'static str,
+    _variant_index: u32,
+    variant: &'static str,
+    len: usize,
+  ) -> Result<Self::SerializeStructVariant> {
+    let scope = self.scope.clone();
+    let x = self.serialize_struct(variant, len)?;
+    Ok(VariantSerializer::new(scope, variant, x))
+  }
+}
+
+macro_rules! not_reachable {
+    ($($name:ident($ty:ty, $lt:lifetime);)*) => {
+        $(fn $name(self, _v: $ty) -> JsResult<$lt> {
+            unreachable!();
+        })*
+    };
+}
```

Comment:
I believe this macro is superfluous, it would be more readable (although repetitive) to spell out those methods explicitly